Provide Kubernetes resource validation to k8s module

willthames opened the pull request:
##### SUMMARY

Use kubernetes-validate to validate Kubernetes resource
definitions against the published schema

Note that `fail_on_error` is compulsory to use `validate` because otherwise the default to obtain some validation would be `validate: {}` which isn't particularly Ansiblish. Better suggestions for a nicer default case are welcome. 

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
k8s

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.7.0.dev0 (devel 0b77262288) last updated 2018/07/27 21:11:54 (GMT +1000)
  config file = None
  configured module search path = [u'/Users/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/will/src/ansible/lib/ansible
  executable location = /Users/will/src/ansible/bin/ansible
  python version = 2.7.14 (default, Mar  9 2018, 23:57:12) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```


##### ADDITIONAL INFORMATION
e.g. with

```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  labels:
    app: kuard
  name: kuard
  namespace: default
spec:
  replicas: 3
  selector:
    matchLabels:
      app: kuard
    unwanted: value
  template:
    metadata:
      labels:
        app: kuard
    spec:
      containers:
      - image: gcr.io/kuar-demo/kuard-amd64:1
        name: kuard
```

and

```
- hosts: localhost

  tasks:
    - k8s:
        definition: "{{ lookup('file', 'kuard-extra-property.yml') }}"
        validate:
          fail_on_error: no
```

the result includes

```
 [WARNING]: resource validation error at spec.selector: Additional properties are not allowed ('unwanted' was unexpected)
```

willthames committed:
Provide Kubernetes resource validation to k8s module

Use kubernetes-validate to validate Kubernetes resource
definitions against the published schema

fabianvf commented:
How does this behave when validating an unknown resource type?

willthames commented:
@fabianvf it fails earlier in the module with `resource.get`. Using `kind: Ostrich`:

```
    "item": "/Users/will/src/kubernetes-validate/examples/kuard-made-up-kind.yaml", 
    "msg": "Failed to find exact match for extensions/v1beta1.Ostrich by [kind, name, singularName, shortNames]"
```

Which is good as my implementation of kubernetes-validate would just fail hard with IOError due to file not found - so that's a bug to address there.

willthames commented:
One thing you might be highlighting is new resource types added through Custom Resource Definitions - I'll have to check how this works with e.g. kubeless on Monday. I might have to warn on missing resources as I'm not sure what else I can do there.

fabianvf commented:
@willthames yeah, I'm talking about resources created through CustomResourceDefinitions and Aggregated API servers. 

Thinking about this more, I think a better approach would be to add the ability to validate to the client and just use it from ansible. That way we can leave the particulars of the library/method used for authentication to the client and just call a validate() method on the resource or something.

In terms of the best way to do client-side validation, there's a proposal floating around for an alternative implementation of the discovery procedure for the dynamic client. Essentially, kubernetes (and by extension openshift) expose an OpenAPI schema through an endpoint on the cluster. If we implemented an alternative discovery that used this endpoint then we could attach the specification for the resource to the resource. This would allow us to support validation for all resources in the cluster, and should be portable across versions. 

Not saying that I would oppose a less effortful implementation of client-side validation to satisfy the short term need, but ideally it would be a fairly seamless transition to the proper full client validation/discovery (in terms of interface to the user / impact on other projects).

willthames commented:
@fabianvf - I'd definitely be keen on a more sophisticated validation through the dynamic client that could cope with Custom Resource Definitions

One short-term option here might be to enumerate the various supported APIs and have a `warn_on_missing_api` or something, and just ignore unknown APIs if that's `False`. 

Now I have a little more experience with jsonschema validation, let me know if I can be of assistance to you in implementing such a feature (or if not, I'll happily reimplement this against that feature)

fabianvf commented on lib/ansible/module_utils/k8s/raw.py:
```diff
@@ -109,6 +128,27 @@ def execute_module(self):
             }
         })
 
+    def validate(self, resource):
```
I think it would make sense to move the actual validation logic to the client, and just call it from here. Other than that lgtm

willthames commented:
Thanks @fabianvf, expect another client PR in the next couple of days (I also have a couple of merges to make in kubernetes-validate before this is ready).

I'll move this to WIP status and update it once the client is ready

---- On Tue, 28 Aug 2018 03:21:07 +1000 notifications@github.com wrote ----

@fabianvf commented on this pull request.

In lib/ansible/module_utils/k8s/raw.py:

> @@ -109,6 +128,27 @@ def execute_module(self):
             }
         })
 
+    def validate(self, resource):
I think it would make sense to move the actual validation logic to the client, and just call it from here. Other than that lgtm

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

mattclay commented on test/integration/targets/k8s/runme.sh:
```diff
@@ -12,14 +12,26 @@ MYTMPDIR=$(mktemp -d 2>/dev/null || mktemp -d -t 'mytmpdir')
 virtualenv --system-site-packages --python "${PYTHON}" "${MYTMPDIR}/openshift-recent"
 source "${MYTMPDIR}/openshift-recent/bin/activate"
-$PYTHON -m pip install 'openshift>=0.7.0'
+$PYTHON -m pip install 'git+https://github.com/willthames/openshift-restclient-python@append_hash'
```
Why are these pip installs done from GH?

willthames replied:
@mattclay this is a WIP PR and those are WIP versions of the modules - once those modules are released I'll update this PR to use the proper pypi versions

mattclay replied:
Thanks for the clarification.

willthames committed:
Additional tests for kubernetes-validate

willthames committed:
Improve k8s error messages on exceptions

Parse the response body for the message rather than returning
a JSON blob

If we've validated and there are warnings, return those too - they
can be more helpful

```
"msg": "Failed to patch object: {\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},
       \"status\":\"Failure\",\"message\":\"[pos 334]: json: decNum: got first char 'h'\",\"code\":500}\n",
```
vs
```
"msg": "Failed to patch object: [pos 334]: json: decNum: got first char 'h'\nresource
        validation error at spec.replicas: 'hello' is not of type u'integer'",
```

willthames committed:
Update versions used

In particular openshift/origin:3.9.0

fabianvf commented:
shipit

gundalow commented:
@willthames Rebase needed

willthames committed:
Add changelog for k8s validate change

willthames commented:
@fabianvf could you sign this one off again please?

fabianvf commented:
shipit

gundalow commented:
@willthames Thank you for the PR
@fabianvf Thank you for the review

New functionlaity merged into `devel` for release in Ansible 2.8
